app/helpers/miq_expression/filter_subst.rb

```diff
@@ -0,0 +1,115 @@
+module FilterSubst
+  # Build an array of expression symbols by recursively traversing the MiqExpression object
+  #   and inserting sequential tokens for each expression part
+  def exp_build_table(exp, quick_search = false)
+    exp_table = []
+    if exp["and"]
+      exp_table.push("(")
+      exp["and"].each do |e|
+        exp_table += exp_build_table(e, quick_search)
+        exp_table.push("AND") unless e == exp["and"].last
+      end
+      exp_table.push(")")
+    elsif exp["or"]
+      exp_table.push("(")
+      exp["or"].each do |e|
+        exp_table += exp_build_table(e, quick_search)
+        exp_table.push("OR") unless e == exp["or"].last
+      end
+      exp_table.push(")")
+    elsif exp["not"]
+      @exp_token ||= 0
+      @exp_token += 1
+      exp[:token] = @exp_token
+      exp_table.push(quick_search ? "NOT" : ["NOT", @exp_token])  # No token if building quick search exp
+      exp_table.push("(") unless ["and", "or"].include?(exp["not"].keys.first)  # No parens if and/or under me
+      exp_table += exp_build_table(exp["not"], quick_search)
+      exp_table.push(")") unless ["and", "or"].include?(exp["not"].keys.first)  # No parens if and/or under me
+    else
+      @exp_token ||= 0
+      @exp_token += 1
+      exp[:token] = @exp_token
+      if exp["???"]                             # Found a new expression part
+        exp_table.push(["???", @exp_token])
+        @edit[@expkey][:exp_token] = @exp_token         # Save the token value for the view
```

Comment:
Can we at least avoid the @edit thing with a helper method...at least then it lives one method that we can kill. :(